### PR TITLE
Add embedded-io, embedded-io-async, remove serial traits.

### PR DESCRIPTION
**Pull request by Dirbaio**
As discussed [previously](https://github.com/rust-embedded/embedded-hal/pull/349#issuecomment-1499465364) and in today's meeting ([chatlog](https://matrix.to/#/!BHcierreUuwCMxVqOf:matrix.org/$K_e5cYsDBFgZ45TN-LRCDH6nVBVyiJI550uoJk8FVGE?via=matrix.org&via=psion.agg.io&via=tchncs.de), [minutes](https://github.com/rust-embedded/wg/blob/master/minutes/2023-07-04.md)), we've decided to not have serial traits, in favor of a more general "io" abstraction.

This PR moves the embedded-io crate to the embedded-hal repo. The crate currently living at https://github.com/embassy-rs/embedded-io, that repo will be archived once the move is complete.

Changes from the embassy repo:
- Removed `defmt` impls, needed for 1.0.
- Split `embedded-io` and `embedded-io-async`. Needed for 1.0, so we don't have to wait for AFIT to be stable.
- Removed async adapters for `tokio` and `futures`. I'm not sure if we want them here or it's better to leave them to community crates. If we do, I'll add them in a future PR.
- Added `embedded-io::{ReadReady, WriteReady}`. They're separate from the `Read`/`Write` traits since they can't be trivially implemented on top of `std::io` traits, so this would present a problem for the adapters. (It's likely implementable with e.g. some weird `poll()/select()` contraption, but that's OS-specific and I'm sure has other tradeoffs, so I'd rather not make the adapters do that by default.)

The crate has been added by parts in several commits to make review easier. 

The `embedded-io` crate has been around for more than 1 year, so it's somewhat "proven". The design is the same as `std::io`, which is more than proven. Highlights of uses of `embedded-io` in the field ([see more](https://crates.io/crates/embedded-io/reverse_dependencies))
- [`embassy-nrf` DMA-powered UART impl](https://github.com/embassy-rs/embassy/blob/ce73c29246d59b913efaafdb38414806541e0511/embassy-nrf/src/buffered_uarte.rs#L572)
- [`embassy-stm32` irq-driven UART](https://github.com/embassy-rs/embassy/blob/ce73c29246d59b913efaafdb38414806541e0511/embassy-stm32/src/usart/buffered.rs#L403C10-L403C11)
- [`embassy-stm32` DMA-powered UART](https://github.com/embassy-rs/embassy/blob/ce73c29246d59b913efaafdb38414806541e0511/embassy-stm32/src/usart/ringbuffered.rs#L223)
- [`embassy-rp` irq-driven UART](https://github.com/embassy-rs/embassy/blob/ce73c29246d59b913efaafdb38414806541e0511/embassy-rp/src/uart/buffered.rs#L571)
- [`embassy-sync` Pipe](https://docs.rs/embassy-sync/0.2.0/embassy_sync/pipe/struct.Pipe.html)
- [`embassy-net` TcpSocket](https://github.com/embassy-rs/embassy/blob/ce73c29246d59b913efaafdb38414806541e0511/embassy-net/src/tcp.rs#L382)
- [`embedded-tls`](https://crates.io/crates/embedded-tls) TLS client, uses embedded-io to work on top of any TCP socket.
- [`reqwless`](https://crates.io/crates/reqwless) HTTP client, uses embedded-io to work on top of any TCP socket or on top of `embedded-tls`.
- [`rust-mqtt`](https://crates.io/crates/rust-mqtt), same.
- [`embedded-svc`](https://crates.io/crates/embedded-svc) abstractions, widely used in the ESP32 ecosystem.
- [`embedded-nal-async`](https://crates.io/crates/embedded-nal-async) models TCP sockets as `Read+Write`.
- [`es-wifi-driver`](https://crates.io/crates/es-wifi-driver), AT commands over uart
- [`esp8266-at-driver`](https://crates.io/crates/esp8266-at-driver), AT commands over uart

### CODE REVIEW

**Review by ryankurte — APPROVED** [review]
approach seems reasonable and examples are solid, lgtm (and thanks for your work on this)!

**Review by therealprof — APPROVED** [review]
LGTM! Thanks a lot.

File: embedded-io/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+    /// If this returns `true`, it's guaranteed that the next call to [`Read::read`] or [`BufRead::fill_buf`] will not block.
+    fn read_ready(&mut self) -> Result<bool, Self::Error>;
+}
+
+/// Get whether a writer is ready.
+///
+/// This allows using a [`Write`] in a nonblocking fashion, i.e. trying to write
+/// only when it is ready.
+pub trait WriteReady: crate::Io {
+    /// Get whether the writer is ready for immediately writeing.
```
Comment (eldruin) [inline embedded-io/src/lib.rs:284]:
```suggestion
    /// Get whether the writer is ready for immediately writing.
```

**Comment by Dirbaio** [comment]
adressed @eldruin 's review comments

File: embedded-io/src/adapters.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }
+}
+
+impl<T: crate::Seek + ?Sized> std::io::Seek for ToStd<T> {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> Result<u64, std::io::Error> {
+        self.inner.seek(pos.into()).map_err(to_io_error)
+    }
+}
+
+fn to_io_error<T: core::fmt::Debug>(err: T) -> std::io::Error {
```
[the remark this answers is not in the document]
Comment (Dirbaio) [reply embedded-io/src/adapters.rs]: renamed!

File: embedded-io/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+/// Base trait for all IO traits.
+///
+/// All IO operations of all traits return the error defined in this trait.
+///
+/// Having a shared trait instead of having every trait define its own
+/// `Error` associated type enforces all impls on the same type use the same error.
+/// This is very convenient when writing generic code, it means you have to
+/// handle a single error type `T::Error`, instead of `<T as Read>::Error` and `<T as Write>::Error`
+/// which might be different types.
+pub trait Io {
```
[the remark this answers is not in the document]
Comment (Dirbaio) [reply embedded-io/src/lib.rs]: renamed!

**Review by eldruin — APPROVED** [review]
Great, thank you again!

**Pull request merged**